fdbclient/GlobalConfig.actor.cpp

```diff
@@ -0,0 +1,182 @@
+		config->_updater = updater(config, dbInfo);
+	}
+}
+
+GlobalConfig& GlobalConfig::globalConfig() {
+	void* res = g_network->global(INetwork::enGlobalConfig);
+	ASSERT(res);
+	return *reinterpret_cast<GlobalConfig*>(res);
+}
+
+Key GlobalConfig::prefixedKey(KeyRef key) {
+	return key.withPrefix(SpecialKeySpace::getModuleRange(SpecialKeySpace::MODULE::GLOBALCONFIG).begin);
+}
+
+const Reference<ConfigValue> GlobalConfig::get(KeyRef name) {
+	auto it = data.find(name);
+	if (it == data.end()) {
+		return Reference<ConfigValue>();
+	}
+	return it->second;
+}
+
+const std::map<KeyRef, Reference<ConfigValue>> GlobalConfig::get(KeyRangeRef range) {
+	std::map<KeyRef, Reference<ConfigValue>> results;
+	for (const auto& [key, value] : data) {
+		if (range.contains(key)) {
+			results[key] = value;
+		}
+	}
+	return results;
+}
+
+Future<Void> GlobalConfig::onInitialized() {
+	return initialized.getFuture();
+}
+
+void GlobalConfig::insert(KeyRef key, ValueRef value) {
+	data.erase(key);
+
+	Arena arena(key.expectedSize() + value.expectedSize());
+	KeyRef stableKey = KeyRef(arena, key);
+	try {
+		std::any any;
+		Tuple t = Tuple::unpack(value);
+		if (t.getType(0) == Tuple::ElementType::UTF8) {
+			any = StringRef(arena, t.getString(0).contents());
+		} else if (t.getType(0) == Tuple::ElementType::INT) {
+			any = t.getInt(0);
+		} else if (t.getType(0) == Tuple::ElementType::FLOAT) {
+			any = t.getFloat(0);
+		} else if (t.getType(0) == Tuple::ElementType::DOUBLE) {
+			any = t.getDouble(0);
+		} else {
+			ASSERT(false);
+		}
+		data[stableKey] = makeReference<ConfigValue>(std::move(arena), std::move(any));
+	} catch (Error& e) {
+		TraceEvent("GlobalConfigTupleError").detail("What", e.what());
+	}
+}
+
+void GlobalConfig::erase(KeyRef key) {
+	erase(KeyRangeRef(key, keyAfter(key)));
+}
+
+void GlobalConfig::erase(KeyRangeRef range) {
+	auto it = data.begin();
+	while (it != data.end()) {
+		if (range.contains(it->first)) {
+			it = data.erase(it);
+		} else {
+			++it;
+		}
+	}
+}
+
+// Updates local copy of global configuration by reading the entire key-range
+// from storage.
+ACTOR Future<Void> GlobalConfig::refresh(GlobalConfig* self) {
+	Transaction tr(self->cx);
+	Standalone<RangeResultRef> result = wait(tr.getRange(globalConfigDataKeys, CLIENT_KNOBS->TOO_MANY));
+	for (const auto& kv : result) {
+		KeyRef systemKey = kv.key.removePrefix(globalConfigKeysPrefix);
+		self->insert(systemKey, kv.value);
+	}
+	return Void();
+}
+
+// Applies updates to the local copy of the global configuration when this
+// process receives an updated history.
+ACTOR Future<Void> GlobalConfig::updater(GlobalConfig* self, Reference<AsyncVar<ClientDBInfo>> dbInfo) {
+	wait(self->refresh(self));
+	self->initialized.send(Void());
+
+	loop {
+		try {
+			wait(dbInfo->onChange());
+
+			auto& history = dbInfo->get().history;
+			if (history.size() == 0) {
+				continue;
+			}
+
+			if (self->lastUpdate < history[0].version) {
```

Comment:
It looks like we never use the mutations from history[0]; either our last update is smaller than history[0] and we can't apply them, or our last update is >= history[0] and we don't need them. Is it worth an optimization to include only the version from the oldest history item?

Reply:
Good point, yes I think this optimization should be easy enough to make, I'll add it.